go/vt/servenv/grpc_codec.go

```diff
@@ -17,52 +17,64 @@ limitations under the License.
 package servenv
 
 import (
-	"fmt"
-
-	// use the original golang/protobuf package we can continue serializing
-	// messages from our dependencies, particularly from etcd
-	"github.com/golang/protobuf/proto" //nolint
-
 	"google.golang.org/grpc/encoding"
+	"google.golang.org/grpc/mem"
+
+	// Guarantee that the built-in proto is called registered before this one
+	// so that it can be replaced.
 	_ "google.golang.org/grpc/encoding/proto" // nolint:revive
 )
 
 // Name is the name registered for the proto compressor.
 const Name = "proto"
 
-type vtprotoCodec struct{}
-
 type vtprotoMessage interface {
-	MarshalVT() ([]byte, error)
+	MarshalToSizedBufferVT(data []byte) (int, error)
 	UnmarshalVT([]byte) error
+	SizeVT() int
 }
 
-func (vtprotoCodec) Marshal(v any) ([]byte, error) {
-	switch v := v.(type) {
-	case vtprotoMessage:
-		return v.MarshalVT()
-	case proto.Message:
-		return proto.Marshal(v)
-	default:
-		return nil, fmt.Errorf("failed to marshal, message is %T, must satisfy the vtprotoMessage interface or want proto.Message", v)
-	}
+type Codec struct {
+	fallback encoding.CodecV2
 }
 
-func (vtprotoCodec) Unmarshal(data []byte, v any) error {
-	switch v := v.(type) {
-	case vtprotoMessage:
-		return v.UnmarshalVT(data)
-	case proto.Message:
-		return proto.Unmarshal(data, v)
-	default:
-		return fmt.Errorf("failed to unmarshal, message is %T, must satisfy the vtprotoMessage interface or want proto.Message", v)
+func (Codec) Name() string { return Name }
+
+var defaultBufferPool = mem.DefaultBufferPool()
+
+func (c *Codec) Marshal(v any) (mem.BufferSlice, error) {
+	if m, ok := v.(vtprotoMessage); ok {
+		size := m.SizeVT()
+		if mem.IsBelowBufferPoolingThreshold(size) {
+			buf := make([]byte, 0, size)
```

Comment:
It seems that we are not updating the length of the buf. We could use `buf := make([]byte, size)` here to fix this.